torch/distributed/device_mesh.py

```diff
@@ -178,42 +128,43 @@ def create_flatten_mesh(
                 )
 
             # Quick return if the flatten mesh has been created before.
-            # TODO: If we decide to restrict flatten initialization once, we should remove
-            # this check and throw an error if the flatten mesh is already created before.
             if (
                 root_mesh in self.root_to_flatten_mapping
                 and mesh_dim_name in self.root_to_flatten_mapping[root_mesh]
             ):
-                return self.root_to_flatten_mapping[root_mesh][mesh_dim_name]
-
-            flattened_mesh_dim_size = math.prod(device_mesh.mesh.size())
-
-            remained_dims_in_root = list(range(root_mesh.mesh.ndim))
-            for flatten_dim_in_root in flatten_dims_in_root:
-                remained_dims_in_root.remove(flatten_dim_in_root)
-
-            pg_ranks_by_dim = root_mesh.mesh.permute(
-                *remained_dims_in_root, *flatten_dims_in_root
-            ).reshape(-1, flattened_mesh_dim_size)
+                if (
+                    device_mesh._layout.coalesce()
+                    == self.flatten_name_to_root_layout[root_mesh][mesh_dim_name]
```

Comment:
Are the meshes here guaranteed to be coalesced? (I can figure this out with more code reading)

Reply:
Great question again. The answer is no. Because if you slicing contiguous dims yes if not no. For example for the mesh (2,2,2) ("dp", "cp", "tp") when you do mesh["cp", "tp"]._flatten() then yes, but if you do mesh["dp", "tp"]._flatten() then no. But the mesh tensor needs to be flattened, that's why we have the change in L155.